psl-core/src/main/java/org/linqs/psl/application/inference/online/OnlineInference.java

```diff
@@ -0,0 +1,309 @@
+        } else if (nextAction.getClass() == DeleteAtom.class) {
+            doDeleteAtom((DeleteAtom)nextAction);
+        } else if (nextAction.getClass() == WriteInferredPredicates.class) {
+            doWriteInferredPredicates((WriteInferredPredicates)nextAction);
+        } else if (nextAction.getClass() == Close.class) {
+            doClose((Close)nextAction);
+        } else {
+            throw new IllegalArgumentException("Action: " + nextAction.getClass().getName() + " Not Supported.");
+        }
+    }
+
+    protected void doAddAtom(AddAtom nextAction) throws IllegalArgumentException {
+        // Resolve Predicate
+        Predicate registeredPredicate = Predicate.get(nextAction.getPredicateName());
+        if (registeredPredicate == null) {
+            throw new IllegalArgumentException("Predicate is not registered: " + nextAction.getPredicateName());
+        }
+
+        switch (nextAction.getPartitionName()) {
+            case "READ":
+                ((OnlineTermStore)termStore).addAtom(registeredPredicate, nextAction.getArguments(), nextAction.getValue(), true);
+                break;
+            case "WRITE":
+                ((OnlineTermStore)termStore).addAtom(registeredPredicate, nextAction.getArguments(), nextAction.getValue(), false);
+                break;
+            default:
+                throw new IllegalArgumentException("Add Atom Partition: " + nextAction.getPartitionName() + "Not Supported");
+        }
+    }
+
+    protected void doDeleteAtom(DeleteAtom nextAction) throws IllegalArgumentException {
+        // Resolve Predicate
+        Predicate registeredPredicate = Predicate.get(nextAction.getPredicateName());
+        if (registeredPredicate == null) {
+            throw new IllegalArgumentException("Predicate is not registered: " + nextAction.getPredicateName());
+        }
+
+        ((OnlineTermStore)termStore).deleteAtom(registeredPredicate, nextAction.getArguments());
+    }
+
+    protected void doUpdateObservation(UpdateObservation nextAction) throws IllegalArgumentException {
+        // Resolve Predicate
+        Predicate registeredPredicate = Predicate.get(nextAction.getPredicateName());
+        if (registeredPredicate == null) {
+            throw new IllegalArgumentException("Predicate is not registered: " + nextAction.getPredicateName());
+        }
+
+        ((OnlineTermStore)termStore).updateAtom(registeredPredicate, nextAction.getArguments(), nextAction.getValue());
+    }
+
+    protected void doClose(Close nextAction) {
+        close = true;
+    }
+
+    protected void doWriteInferredPredicates(WriteInferredPredicates nextAction) {
+        long start;
+        long stop;
+        // Activate the atoms that were added
+        log.trace("Partial Grounding Start");
+        start = System.currentTimeMillis();
+        ArrayList<GroundRule> groundRules = ((OnlineAtomManager)atomManager).activateAtoms(rules, (OnlineTermStore) termStore);
```

Comment:
This is a big one.
Here, in this array list, results are being pooled (collected).
We need to avoid any pooling that scales with the number of terms in streaming/online applications.
So the OnlineAtomManager needs to be modified so it returns query results as an iterator (like the classic grounding).